### PR TITLE
[XGBoost] Add GPU support 

### PR DESCRIPTION
**Pull request by tylerjthomas9**
This PR is a WIP. I am trying to build XGBoost with GPU support (https://github.com/dmlc/XGBoost.jl/issues/134). 

Currently, the `x86_64-linux-gnu` cuda build is working. I am struggling to get it to also build with NCCL, but that is not a necessity. 

The `x86_64-w64-mingw` cuda build is not working. Cmake throws the following error:

```
CMake Error at /usr/share/cmake/Modules/CMakeTestCUDACompiler.cmake:56 (message):
[19:10:08]   The CUDA compiler
[19:10:08] 
[19:10:08]     "/workspace/destdir/cuda/bin/nvcc.exe"
[19:10:08] 
[19:10:08]   is not able to compile a simple test program.
[19:10:08] 
[19:10:08]   It fails with the following output:
[19:10:08] 
[19:10:08]     Change Dir: /workspace/srcdir/xgboost/build/CMakeFiles/CMakeTmp
[19:10:08]     
[19:10:08]     Run Build Command(s):/usr/bin/make -f Makefile cmTC_e97c7/fast && /usr/bin/make  -f CMakeFiles/cmTC_e97c7.dir/build.make CMakeFiles/cmTC_e97c7.dir/build
[19:10:08]     make[1]: Entering directory '/workspace/srcdir/xgboost/build/CMakeFiles/CMakeTmp'
[19:10:08]     Building CUDA object CMakeFiles/cmTC_e97c7.dir/main.cu.obj
[19:10:08]     /workspace/destdir/cuda/bin/nvcc.exe      -c /workspace/srcdir/xgboost/build/CMakeFiles/CMakeTmp/main.cu -o CMakeFiles/cmTC_e97c7.dir/main.cu.obj
[19:10:08]     /workspace/destdir/cuda/bin/nvcc.exe: line 1: MZÿÿ¸@º´: not found
[19:10:08]     /workspace/destdir/cuda/bin/nvcc.exe: line 1: syntax error: unexpected end of file (expecting ")")
[19:10:08]     make[1]: *** [CMakeFiles/cmTC_e97c7.dir/build.make:78: CMakeFiles/cmTC_e97c7.dir/main.cu.obj] Error 2
[19:10:08]     make[1]: Leaving directory '/workspace/srcdir/xgboost/build/CMakeFiles/CMakeTmp'
[19:10:08]     make: *** [Makefile:127: cmTC_e97c7/fast] Error 2

```

### CODE REVIEW

**Comment by giordano** [comment]
https://docs.binarybuilder.org/stable/troubleshooting/#Running-foreign-executables.  See also #5863 for basically the same problem.  Side note, it isn't clear to me what happens if one doesn't have CUDA.  Would this even work?  If not, this doesn't sound like a good idea at all.

**Comment by imciner2** [comment]
I don't think nvcc/CUDA supports cross compilation, so if you want to package a GPU backend you can only do it on systems that build natively (i.e. only linux 64).

**Comment by tylerjthomas9** [comment]
> Side note, it isn't clear to me what happens if one doesn't have CUDA. Would this even work? If not, this doesn't sound like a good idea at all.

I just updated the `build_tarballs.jl` to have CUDA and non CUDA versions of the linux builds. Would this fix the potential issue? Or, would it be better to separate the CUDA and non-cuda scripts? Similar to how SuiteSparse handles it. 

> I don't think nvcc/CUDA supports cross compilation, so if you want to package a GPU backend you can only do it on systems that build natively (i.e. only linux 64).

Ah, I think you are right.

**Comment by ViralBShah** [comment]
Good to have @maleadt review as well.

File: X/XGBoost/build_tarballs.jl [hunk 1]
```diff
@@ -37,9 +57,21 @@ fi
 install_license LICENSE
 """
 
+
 # These are the platforms we will build for by default, unless further
 # platforms are passed in on the command line
 platforms = expand_cxxstring_abis(supported_platforms())
+cuda_platforms = expand_cxxstring_abis(Platform("x86_64", "linux"))
+
+cuda_versions = [v"11.0", v"11.8"]
```
Comment (maleadt) [inline X/XGBoost/build_tarballs.jl:66]: Do you need a separate build for 11.8? If XGBoost doesn't have any specific code targeting it, its minimum requirement of 11.0 should be enough.

**Comment by imciner2** [comment]
I'll defer to @maleadt for a more thorough CUDA review (since I am still trying to wrap my head around how to do library building with the new CUDA runtime jll), but two things I see are:

1) Why are you building for three different CUDA 11 versions? My understanding (https://docs.nvidia.com/deploy/cuda-compatibility/index.html#default-to-minor-version) is that if you build for CUDA 11 you should be able to run the library on other CUDA 11 runtimes (they do have the caveat about limited feature sets, so is this using anything specifically from one of those other CUDA versions that makes this needed?)
2) I think you also need to add the runtime dependency on `CUDA_Runtime_jll` to ensure the appropriate libraries are installed on the user machine, so this should be added to the CUDA dependencies list:
```
RuntimeDependency(PackageSpec(name="CUDA_Runtime_jll"))
```

**Comment by tylerjthomas9** [comment]
> 1. Why are you building for three different CUDA 11 versions? My understanding (https://docs.nvidia.com/deploy/cuda-compatibility/index.html#default-to-minor-version) is that if you build for CUDA 11 you should be able to run the library on other CUDA 11 runtimes (they do have the caveat about limited feature sets, so is this using anything specifically from one of those other CUDA versions that makes this needed?)

It is not needed. I have removed the extra CUDA builds.

> 2. I think you also need to add the runtime dependency on `CUDA_Runtime_jll` to ensure the appropriate libraries are installed on the user machine, so this should be added to the CUDA dependencies l

Done. 

Thank you for the help.

**Comment by giordano** [comment]
What are the dependencies of that library?

**Comment by tylerjthomas9** [comment]
The R Library requires 'data.table', 'jsonlite'. I stumbled across [this issue](https://github.com/dmlc/XGBoost.jl/issues/70), where just replacing the dll seemed to give drop-in compatibility for gpu training on windows.

**Comment by giordano** [comment]
I was referring to the binary library xgboost.dll. I'm not sure how R is related to this

**Comment by tylerjthomas9** [comment]
> I was referring to the binary library xgboost.dll. I'm not sure how R is related to this

Oh, I referenced a R build as a place to get the dll file, so I thought you were talking about that. I did not realize that you could look at dll file dependencies. Here is what I get:

```
xgboost\src>ldd xgboost.dll
        ntdll.dll => /c/WINDOWS/SYSTEM32/ntdll.dll (0x7ffd93a90000)
        KERNEL32.DLL => /c/WINDOWS/System32/KERNEL32.DLL (0x7ffd92780000)
        KERNELBASE.dll => /c/WINDOWS/System32/KERNELBASE.dll (0x7ffd91190000)
        msvcrt.dll => /c/WINDOWS/System32/msvcrt.dll (0x7ffd92080000)
        xgboost.dll => /c/Users/TylerThomas/Downloads/xgboost/src/xgboost.dll (0x7ffcc5340000)
        ws2_32.dll => /c/Windows/System32/ws2_32.dll (0x7ffd93210000)
        rpcrt4.dll => /c/Windows/System32/rpcrt4.dll (0x7ffd924b0000)
        vcomp140.dll => /c/Windows/System32/vcomp140.dll (0x7ffd76010000)
        user32.dll => /c/Windows/System32/user32.dll (0x7ffd93890000)
        win32u.dll => /c/Windows/System32/win32u.dll (0x7ffd91530000)
        gdi32.dll => /c/Windows/System32/gdi32.dll (0x7ffd92480000)
        gdi32full.dll => /c/Windows/System32/gdi32full.dll (0x7ffd91560000)
        msvcp_win.dll => /c/Windows/System32/msvcp_win.dll (0x7ffd917f0000)
        ucrtbase.dll => /c/Windows/System32/ucrtbase.dll (0x7ffd90fb0000)
        R.dll => not found
```

File: X/XGBoost/build_tarballs.jl (hunk 1 above)
Comment (tylerjthomas9) [reply X/XGBoost/build_tarballs.jl]: It is not needed. I have removed the extra CUDA builds.

**Comment by tylerjthomas9** [comment]
Is it a bad idea to pull windows binaries from the Python wheels? If so, we can just add the `x86_64-linux-gnu` + cuda support right now, and expand support down the road.

**Comment by giordano** [comment]
Do I understand correctly that this shouldn't be changing much for the existing platforms but only adding a CUDA build when CUDA is available (on x86_64 Linux)?  If so, that'd be good for me.  @maleadt could you please have another look at the CUDA-related changes?

**Comment by ExpandingMan** [comment]
Thanks a lot @tylerjthomas9 for this contribution.  I was planning on adding GPU support to XGBoost.jl but was having a lot of trouble getting BinaryBuilder to compile with CUDA.

Once this is merged I will get to work adding GPU support to the wrapper!

**Pull request merged**